### ISSUE TITLE
Error linting file w/o icontract

### ISSUE THREAD

**Issue by renefritze** [Parquery/pyicontract-lint#27]
I'm trying out icontract and this linter. I get a fatal error in [this file](https://github.com/pymor/pymor/blob/contracts_type-hints/src/pymor/algorithms/rules.py)

[pip freeze output](https://github.com/Parquery/pyicontract-lint/files/5429910/freeze.txt)
```
pyicontract-lint ./src/pymor/algorithms/rules.py

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 132, in raise_if_nothing_inferred
    yield next(generator)
StopIteration: {'node': <Attribute.instancemethod l.209 at 0x7f2f85040350>, 'context': <astroid.context.InferenceContext object at 0x7f2f850a8470>}

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/pyicontract-lint", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/main.py", line 70, in main
    return _main(args=args, stream=sys.stdout)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/main.py", line 49, in _main
    errors = icontract_lint.check_paths(paths=args.paths)
  File "/usr/local/lib/python3.7/site-packages/icontract/_checkers.py", line 396, in wrapper
    result = func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 576, in check_paths
    errs.extend(check_recursively(path=pth))
  File "/usr/local/lib/python3.7/site-packages/icontract/_checkers.py", line 396, in wrapper
    result = func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 556, in check_recursively
    errs.extend(check_file(pth))
  File "/usr/local/lib/python3.7/site-packages/icontract/_checkers.py", line 396, in wrapper
    result = func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 541, in check_file
    lint_visitor.visit(node=tree)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 109, in visit
    return func(node)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 114, in visit_generic
    self.visit(child)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 109, in visit
    return func(node)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 483, in visit_ClassDef
    self.visit(child)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 109, in visit
    return func(node)
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 397, in visit_FunctionDef
    decorators = [self._infer_decorator(node=decorator_node) for decorator_node in node.decorators.nodes]
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 397, in <listcomp>
    decorators = [self._infer_decorator(node=decorator_node) for decorator_node in node.decorators.nodes]
  File "/usr/local/lib/python3.7/site-packages/icontract_lint/__init__.py", line 364, in _infer_decorator
    decorator = next(node.infer())
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 137, in raise_if_nothing_inferred
    raise exceptions.InferenceError(**error.args[0])
astroid.exceptions.InferenceError: Inference failed for <Attribute.instancemethod l.209 at 0x7f2f85040350>.
```

It's reproducable on mybinder if you click `new -> terminal` and run `pyicontract-lint /pymor/src/pymor/algorithms/rules.py`
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/pymor/pymor/contracts_type-hints)

**Comment by mristin** [comment]
Hi @renefritze ,
Thanks for spotting the bug. I am looking into the issue right now.

**Comment by mristin** [comment]
Hi @renefritze ,
I think I found the bug. It is caused by Astroid having trouble with recursive decorators (if I understand correctly what is going on in the code in `RuleTable` given in [your example at line 209][1])).

Would you mind checking out the code from the pull request https://github.com/Parquery/pyicontract-lint/pull/28 and testing whether there are any other problems?

[1]: https://github.com/pymor/pymor/blob/15896398f49e673c885d54d5f648bb5131efca9b/src/pymor/algorithms/rules.py#L209

**Comment by renefritze** [comment]
I can confirm your fix! Thanks a lot. I ran the linter over our whole package in fact now without further issue :+1:

**Comment by mristin** [comment]
Hi @renefritze ,
I bumped the patch version to 2.1.1 with the fix related to this issue (#30, https://pypi.org/project/pyicontract-lint/). Please let me know if there are any other problems.

Btw., if you get some time -- could you please start the project and encourage other people to do so? I suppose many people are reluctant to adopt the tool given its low star rating (only meager 4 stars for now).
